pinot-controller/src/main/java/org/apache/pinot/controller/recommender/data/writer/CsvWriter.java

```diff
@@ -38,6 +38,13 @@ protected String generateRow(DataGenerator generator) {
     return StringUtils.join(values, ",");
   }
 
+  @Override
+  protected void preprocess(java.io.FileWriter writer)
+      throws Exception {
+    final String headers = StringUtils.join(_spec.getGenerator().nextRow().keySet(), ",");
```

Comment:
I think this will be wrong since we'll end up calling nextRow() for each file, instead we should do it just once. Maybe override init() in this class and store the headers?

Reply:
It won't be wrong, as we can generate as many rows as we want but yes it makes sense to move it to init() to avoid the rework for each file.